core/src/main/java/org/infinispan/protostream/impl/StringUtil.java

```diff
@@ -0,0 +1,155 @@
+         }
+
+         // Coder optimisations not possible
+         INSTANCE = new Utf8Helper() {
+            @Override
+            public boolean providesLatin1Bytes() {
+               return false;
+            }
+
+            @Override
+            public byte[] getBytes(String s) {
+               return s.getBytes(StandardCharsets.UTF_8);
+            }
+         };
+      }
+      return INSTANCE;
+   }
+
+   private static Utf8Helper createLambdaHelper() throws Throwable {
+      Class<MethodHandles.Lookup> lookupClass = MethodHandles.Lookup.class;
+      Field implLookup = lookupClass.getDeclaredField("IMPL_LOOKUP");
+
+      Field theUnsafeField = Unsafe.class.getDeclaredField("theUnsafe");
+      theUnsafeField.setAccessible(true);
+      Unsafe unsafe = (Unsafe) theUnsafeField.get(null);
+      long fieldOffset = unsafe.staticFieldOffset(implLookup);
+      MethodHandles.Lookup trustedLookup = (MethodHandles.Lookup) unsafe.getObject(lookupClass, fieldOffset);
+      MethodHandle constructor = trustedLookup.findConstructor(
+            MethodHandles.Lookup.class,
+            methodType(void.class, Class.class, Class.class, int.class)
+      );
+
+      MethodHandles.Lookup lookup = (MethodHandles.Lookup) constructor.invoke(String.class, null, -1);
+      MethodHandle coder = lookup.findSpecial(
+            String.class,
+            "coder",
+            methodType(byte.class),
+            String.class
+      );
+      CallSite applyAsInt = LambdaMetafactory.metafactory(
+            lookup,
+            "applyAsInt",
+            methodType(ToIntFunction.class),
+            methodType(int.class, Object.class),
+            coder,
+            MethodType.methodType(byte.class, String.class)
+      );
+      ToIntFunction<String> coderFn = (ToIntFunction<String>) applyAsInt.getTarget().invokeExact();
+
+      MethodHandle value = lookup.findSpecial(
+            String.class,
+            "value",
+            methodType(byte[].class),
+            String.class
+      );
+      CallSite apply = LambdaMetafactory.metafactory(
+            lookup,
+            "apply",
+            methodType(Function.class),
+            methodType(Object.class, Object.class),
+            value,
+            methodType(byte[].class, String.class)
+      );
+      Function<String, byte[]> valueFn = (Function<String, byte[]>) apply.getTarget().invokeExact();
+
+      return new Utf8Helper() {
+         @Override
+         public boolean providesLatin1Bytes() {
+            return true;
+         }
+
+         @Override
+         public byte[] getBytes(String s) {
+            if (coderFn.applyAsInt(s) != LATIN1)
+               return s.getBytes(StandardCharsets.UTF_8);
+
+            return valueFn.apply(s);
+         }
+      };
+   }
+
+   private static Utf8Helper createFieldHelper() throws Throwable {
+      Method privateMethod = String.class.getDeclaredMethod("coder");
```

Comment:
I suggest we use `isLatin1` instead. We aren't covering the case of when compact strings aren't used with this.